User Schema Docs (#661)
DevOps-MBSE/AaC: docs/source/project_documentation/user_guide/schema_definitions.md

# Schema Definitions in AaC

AaC employs a set of specialized definitions, all of which are defined with the root-level key `schema`, to establish data structures within the AaC Domain-Specific Language (DSL). These `schema`-defined data structures encompass various forms, such as internal and external message formats, inputs and outputs for modeled behaviors, custom structures for plugin developers or user libraries, and other essential data constructs needed for your team's modeling activities.

In the AaC DSL, every definition has its structure defined by a `schema` definition, which permits users to tailor the language to their team's requirements and even craft new DSLs or extensions within AaC.

Users can define these extensions locally in their user-defined AaC files, or they can leverage [plugins](../dev_guide/plugin_dev_guide.md) and [user libraries](./user_library.md) for common structures and functionalities shared across teams and repositories.

## Utilizing Schema Definitions

> Expanding on this, `schema` definitions serve multiple roles in the AaC DSL. We present specific examples below for various common use cases, including defining instances of custom data structures via `schema` definitions, registering unique root keys for new data structures, and crafting message interfaces for internal and external system messages.

Reviewer: ```suggestion
`Schema` definitions serve multiple roles in the AaC DSL. Some examples are provided below for common use cases: custom data structures, registering unique root keys, and crafting message interfaces.
```

> Expanding on this, `schema` definitions serve multiple roles in the AaC DSL. We present specific examples below for various common use cases, including defining instances of custom data structures via `schema` definitions, registering unique root keys for new data structures, and crafting message interfaces for internal and external system messages.

Author: Done. 👍

> ### Crafting Custom Data Structures

Reviewer: ```suggestion
### Crafting Message Interfaces
```

> One prevalent application of `schema` definitions is crafting and defining interface message structures for communication within your system or between system components. These interface messages define payloads exchanged with external systems, software applications, external actors like users, input for plugins, and intricate data inputs for validation purposes. The way you leverage a `schema` definition depends on the context; if used in the input of a `model` definition's `behavior:input` or `behavior:output`, the `schema` acts as an interface message, documenting the data structure flowing into or out of a component within your modeled system. This is particularly useful for documenting software application behavior inputs or outputs.

Reviewer: ```suggestion
Interface message structures provide structures for communication within your system or between system components. Using `schemas` one can define payloads exchanged with external systems, software applications, external actors, input for plugins, and intricate data inputs for validation purposes. Leveraging a `schema` definition as the `input` of a `model` definition's `behavior:input` or `behavior:output`, the `schema` acts as an interface message, documenting the data structure flowing into or out of a component within your modeled system. A useful way of documenting software application behavior inputs or outputs.
```

> For instance, in our Alarm Clock model example, the `behavior` entry defines the input and behavior of our alarm clock model. In the following example, the behavior contains an input named `Timestamp`, which is a `schema`-defined data structure representing information needed to set the alarm.

Reviewer: ```suggestion
In our Alarm Clock model example, the `behavior` entry defines the input and behavior of our Alarm Clock model. The `behavior` contains an `input` named `Timestamp`, which is a `schema`-defined data structure representing information needed to set the alarm.
```

```{eval-rst}
.. literalinclude:: ../../../../python/model/alarm_clock/alarm_clock.yaml
    :language: yaml
    :emphasize-lines: 17
    :lines: 6-28
```

The `Timestamp` `schema` definition below outlines the fields required to define an exact alarm time.

```{eval-rst}
.. literalinclude:: ../../../../python/model/alarm_clock/structures.yaml
    :language: yaml
    :emphasize-lines: 2
    :lines: 12-28
```

While this example is simplified and not representative of a software system, the same approach applies to defining user requests, events from message queues, Swift XML messages in financial applications, or any other input/output data structure within your modeled system.

### Crafting Custom Data Structures

> Given AaC's self-defining nature, users can employ the `schema` definition to craft their own data structures. In the context of our Alarm Clock example, users simply need to define a new data structure using the `schema` root key. `Timestamp` serves as an example of a defined data structure that can be referenced by name, as demonstrated in the Alarm Clock model's `behavior`, or used as a component in more intricate data structures.

Reviewer: ```suggestion
Given AaC's self-defining nature, users can employ the `schema` definition to craft their own data structures by using the `schema` root key. `Timestamp` serves as an example of a defined data structure with a `schema` root key that can be referenced by name, as demonstrated in the Alarm Clock model's `behavior`, or used as a component in more intricate data structures.
```

```{eval-rst}
.. literalinclude:: ../../../../python/model/alarm_clock/structures.yaml
    :language: yaml
    :emphasize-lines: 2
    :lines: 12-28
```

> ### Constructing Complex Data Structures via Composition

Reviewer: ```suggestion
#### Constructing Complex Data Structures via Composition
```
